src/app/clusters/scenes/SceneTableImpl.cpp

```diff
@@ -0,0 +1,467 @@
+
+using SceneTableEntry = SceneTableImpl::SceneTableEntry;
+using SceneStorageId  = SceneTableImpl::SceneStorageId;
+using SceneData       = SceneTableImpl::SceneData;
+
+struct FabricList : public CommonPersistentData::FabricList
+{
+    CHIP_ERROR UpdateKey(StorageKeyName & key) override
+    {
+        key = DefaultStorageKeyAllocator::SceneFabricList();
+        return CHIP_NO_ERROR;
+    }
+};
+
+static constexpr size_t kPersistentBufferMax = 128;
+
+/**
+ * @brief Linked list of all scenes in a fabric, stored in persistent memory
+ *
+ * FabricSceneData is an access to a linked list of scenes
+ */
+struct FabricSceneData : public PersistentData<kPersistentBufferMax>
+{
+    // static constexpr TLV::Tag TagTableID() { return TLV::ContextTag(1); }
+    static constexpr TLV::Tag TagSceneCount() { return TLV::ContextTag(1); }
+    static constexpr TLV::Tag TagNext() { return TLV::ContextTag(2); }
+
+    chip::FabricIndex fabric_index = kUndefinedFabricIndex;
+    SceneTableImpl::SceneStorageId first_scene;
+    uint16_t scene_count = 0;
+    FabricIndex next     = kUndefinedFabricIndex;
+
+    FabricSceneData() = default;
+    FabricSceneData(FabricIndex fabric) : fabric_index(fabric) {}
+
+    CHIP_ERROR UpdateKey(StorageKeyName & key) override
+    {
+        VerifyOrReturnError(kUndefinedFabricIndex != fabric_index, CHIP_ERROR_INVALID_FABRIC_INDEX);
+        key = DefaultStorageKeyAllocator::FabricScenesKey(fabric_index);
+        return CHIP_NO_ERROR;
+    }
+
+    void Clear() override
+    {
+        first_scene.clear();
+        scene_count = 0;
+        next        = kUndefinedFabricIndex;
+    }
+
+    CHIP_ERROR Serialize(TLV::TLVWriter & writer) const override
+    {
+        TLV::TLVType container;
+        ReturnErrorOnFailure(writer.StartContainer(TLV::AnonymousTag(), TLV::kTLVType_Structure, container));
+
+        ReturnErrorOnFailure(first_scene.Serialize(writer));
+        ReturnErrorOnFailure(writer.Put(TagSceneCount(), static_cast<uint16_t>(scene_count)));
+        ReturnErrorOnFailure(writer.Put(TagNext(), static_cast<uint16_t>(next)));
+
+        return writer.EndContainer(container);
+    }
+
+    CHIP_ERROR Deserialize(TLV::TLVReader & reader) override
+    {
+        ReturnErrorOnFailure(reader.Next(TLV::AnonymousTag()));
+        VerifyOrReturnError(TLV::kTLVType_Structure == reader.GetType(), CHIP_ERROR_INTERNAL);
+
+        TLV::TLVType container;
+        ReturnErrorOnFailure(reader.EnterContainer(container));
+
+        ReturnErrorOnFailure(first_scene.Deserialize(reader));
+        ReturnErrorOnFailure(reader.Next(TagSceneCount()));
+        ReturnErrorOnFailure(reader.Get(scene_count));
+        ReturnErrorOnFailure(reader.Next(TagNext()));
+        ReturnErrorOnFailure(reader.Get(next));
+
+        return reader.ExitContainer(container);
+    }
+
+    // Register the fabric in the fabrics' linked-list
+    CHIP_ERROR Register(PersistentStorageDelegate * storage)
+    {
+        FabricList fabric_list;
```

Comment:
FabricsHavingScenesList type or similar variable name may help make it less ambiguous that this is not at all related to the top-level FabricTable.